#Jay Agyemang - Final Project Info 3340


# **Table of Contents**



>>[Project Details](#scrollTo=pYiNFGHRHKqD)

>>[Data Cleaning Process](#scrollTo=BBDC6EwYALK8)

>[Solution to NBA Draftkings Problem](#scrollTo=f93Ns0owCjME)

>>[Problem 1 Selecting An Optimal Line up With Historical Information:](#scrollTo=3b3Imwhc82QU)

>>[Problem 2 Selecting An Optimal Line up with actual peformance on Feb 22:](#scrollTo=08PmGwTfEOud)

>>[Data](#scrollTo=aPQ5E09mDU2B)

>>[Problem 2](#scrollTo=YzWbVisUOIDh)

>>[Problem 1](#scrollTo=4yJTR9jvDi0v)

>>[Final Report](#scrollTo=A2MR_LWLlDV0)

>>[Appendix 1 - Ask your favorite GAI what it would do](#scrollTo=l8PSDagQcCIU)

>>[Appendix 2 - Any documentation of the use of GAI](#scrollTo=3RZa8fPlcEwB)



## 1. Project Details <a id="1"></a>

**Description**

Who wants to win at Daily Fantasy Sports on Draftkings? Everybody? Anybody? Well, let’s be honest, this project isn’t going to make you win, but it is a fun optimization problem! And that’s really the point in this class.

Imagine that we were interested in joining one of the Draftkings’ National Basketball Association (NBA) contests on Feb 22, 2024. There were 12 real NBA games played amongst 24 teams. The fantasy contest involves selecting a collection of players (a lineup) in these 12 games and you “play” against other people that (potentially) selected different lineups. The winner of the contest is the person whose team accumulated the most fantasy points. The fantasy points are accumulated according to a set of rules (see Figure 1) that are related to how the players perform in their real-life games. Selecting a lineup is subject to a fictitious salary cap and several additional constraints that are outlined below. The goal of this project is to select an “optimal” lineup. What is “optimal?” That is for you to decide!  

**Data**

Here are the data that you need to solve this problem. I will provide a data dictionary on Canvas related to each data set describing the column names.
•	DKSalaries.csv – Data from Draftkings website including player names, teams, salary, position, and average points per game.
•	Actual game data from Feb 22, 2024
o	XXX.csv – Data from each of the 12 games on Feb 22 related to the players.
•	Season Statistics
o	PlayerStats.csv – Data from the NBA season (October 24, 2023 through Feb 21, 2024)

**Objectives**
1.	Use python to apply optimization models and solution techniques learned in class to solve a real-world problem.
2.	You will have to do some exploratory data analysis using the historical data to carefully consider your objective function coefficients.
3. You will solve at least two optimization problems for this project:

  a. One problem will involve selecting a lineup based on historical information. Tell me what your strategy was in choosing players, and how it would outperform the naïve strategy of picking just based on average daily points.  Perhaps older players will play better than they averaged because these players had a lot of rest?  Maybe players do better at home than on the road?  Maybe players do better if they are playing against a bad team, or worse if they are playing against a good team? Maybe you want to pick people who are on the same team, because if one player is scoring, his teammate might be getting the assist?  If you don’t know anything about basketball, use Google (or your favorite AI)!
  
  b.	The second problem will involve selecting the optimal lineup based on how the players performed during their actual games (i.e., the “cheat code” if you had it prior to selecting your lineup.
4.	Write a report summarizing your results.

**Constraints**
1.	The fantasy team must have eight players.
2.	The salary of your eight players may not exceed $50,000.
3.	You need players from at least two different games.
4.	Your 8-player fantasy team roster should include one player in each of these positions:

  a.	Point Guard (PG)

  b.	Shooting Guard (SG)

  c.	Small Forward (SF)

  d.	Power Forward (PF)
  
  e.	Center (C)

  f.	Guard (either PG or SG)

  g.	Forward (either SF or PF)

  h.	Utility (any of the 5 positions)


##Data Cleaning

Consolidated the contents of all three CSV files into a single Excel file for ease of management.
Rectified player names for consistency, such as transforming "Dennis Schröder" into "Dennis Schröder."
Incorporated Salary, Roster Position, and Game Info columns from both the Feb22NBAStats and PlayerStats files into one consolidated file, facilitating streamlined analysis. Additionally, I introduced columns named "Double-Double" and "Triple Double" in the Feb22NBAStats file. This was achieved through the utilization of VLOOKUPs to integrate existing data between files and IF statements to generate the new columns.
Eliminated players lacking salary data from both the Feb22NBAStats and PlayerStats files to ensure data integrity and consistency.









# **Solution to NBA Draftkings Problem**


## **Problem 1 Selecting An Optimal Line up With Historical Information:**

After adding three additonal constriants of playing at least 1000 minutes, not being a player on the Detroit Pistons or Washington Wizards, and having a field goal percentage of 45% or greater this to the original constriants listed above.

Here is my optimal Lineup:

Point Guard (PG):**Scottie Barnes**

Shooting Guard (SG):**Benedict Mathurin**

Small Forward (SF):**DeMar DeRozan**

Power Forward (PF):**Paolo Banchero**
  
Center (C): **Jonas Valančiūnas**

Guard (either PG or SG):**Cason Wallace**

Forward (either SF or PF):**Obi Toppin**

Utility (any of the 5 positions):**Chet Holmgren**

---
**Report Total Average Points per Game**

The total average points per game for players across the league was **8.97**  while the total average points for my optimal lineup was **15.96** . My line up outpeformed the league average.



---

**Report Total Salary**

The total salary for my optimal line with the addition of the three new constriants is  **$49,900**.

 This is an indication that my line is great line up since I spent just 1000 dollars under the allowed salary cap **$50,000**



## **Problem 2 Selecting An Optimal Line up with actual peformance on Feb 22:**
This problem was solved by using pyomo to optmize to the original constraints listed above with the actual peformance of players on Feb 22

Here is my optimal Lineup:

Point Guard (PG):**Luka Dončić**

Shooting Guard (SG):**Derrick White**

Small Forward (SF):**Herbet Jones**

Power Forward (PF):**Kyle Kuzma**
  
Center (C): **Jonas Valančiūnas**

Guard (either PG or SG):**T.J. McConnell**

Forward (either SF or PF):**Ausar Thompson**

Utility (any of the 5 positions):**Gary Trent Jr.**

---
**Report Draftkings Points for Actual Games**

The Drafkings Points for this line up based on the actual games played on Feb 22 is : **386.75**



---

**Report Total Salary**

The total salary for my optimal line with the addition of the three new constriants is  **$50,000**.

 This is an indication that my line is great line up since I spent the maxiumum that I could on players since the salary cap is $50,000 and this problem was optimized of the actual stats when these players played on Feb 22


---
** Report the total avg pts per game**

The total average points for every game of Feb 22 was **13.62** while the average total points for this lineup was **23.62**. The lineup selected on Feb 22 out performed the total average points from the league that by a significant margin.




##Data

In [ ]:
!pip install pyomo
!apt-get install -y -qq glpk-utils

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 26.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 4.8 MB/s eta 0:00:00
Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 121752 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.10.1+dfsg-4build1_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.10.1+dfsg-4build1) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...
Unpacking libamd2:amd64 (1:5.10.1+dfsg-4build1) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.10.1+dfsg-4build1) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_5.0-1_amd64.deb ...
Unpacking libglpk40:amd64 (5.0-1) ...
Selecting previously unselected pac

In [ ]:
import pyomo.environ as pe
import pandas as pd

In [ ]:
import os
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!jupyter nbconvert --to html /content/FinalProjectNBA.ipynb

[NbConvertApp] WARNING | pattern '/content/FinalProjectNBA.ipynb' matched no files
This application is used to convert notebook files (*.ipynb)
        to various other formats.


Options
The options below are convenience aliases to configurable class-options,
as listed in the "Equivalent to" description-line of the aliases.
To see all configurable class-options for some <cmd>, use:
    <cmd> --help-all

--debug
    set log level to logging.DEBUG (maximize logging output)
    Equivalent to: [--Application.log_level=10]
--show-config
    Show the application's configuration (human-readable format)
    Equivalent to: [--Application.show_config=True]
--show-config-json
    Show the application's configuration (json format)
    Equivalent to: [--Application.show_config_json=True]
--generate-config
    generate default config file
    Equivalent to: [--JupyterApp.generate_config=True]
-y
    Answer yes to any questions instead of prompting.
    Equivalent to: [--JupyterApp.answer_yes=True]


In [ ]:
# Read the Excel file
file_path = '/content/drive/My Drive/Colab Notebooks/NBAProject1.xlsx'

# Read the Excel file
data = pd.read_excel(file_path, sheet_name=None)

# Extract data from different sheets
dk_salaries = data['DKSalaries']
player_stats = data['PlayerStats']
feb22_stats = data['Feb22NBAStats']

In [ ]:
dk_salaries

Position                           Name + ID                     Name  \
0         PG              Luka Doncic (32706645)              Luka Doncic   
1          C             Nikola Jokic (32706648)             Nikola Jokic   
2          C              Joel Embiid (32706650)              Joel Embiid   
3         PG  Shai Gilgeous-Alexander (32706652)  Shai Gilgeous-Alexander   
4         PF             Kevin Durant (32706655)             Kevin Durant   
..       ...                                 ...                      ...   
337        C             Jock Landale (32707714)             Jock Landale   
338        C         Boban Marjanovic (32707716)         Boban Marjanovic   
339       PG            Nate Williams (32707718)            Nate Williams   
340       SG       Reggie Bullock Jr. (32707721)       Reggie Bullock Jr.   
341        C             Steven Adams (32707724)             Steven Adams   

           ID Roster Position  Salary                      Game Info  \
0    32706645       PG/G/UTIL   12100  PHX@DAL 02/22/2024 07:30PM ET   
1    32706648          C/UTIL   11600  WAS@DEN 02/22/2024 09:00PM ET   
2    32706650          C/UTIL   11000  NYK@PHI 02/22/2024 07:00PM ET   
3    32706652       PG/G/UTIL   10500  LAC@OKC 02/22/2024 08:00PM ET   
4    32706655       PF/F/UTIL   10000  PHX@DAL 02/22/2024 07:30PM ET   
..        ...             ...     ...                            ...   
337  32707714          C/UTIL    3000  HOU@NOP 02/22/2024 08:00PM ET   
338  32707716          C/UTIL    3000  HOU@NOP 02/22/2024 08:00PM ET   
339  32707718       PG/G/UTIL    3000  HOU@NOP 02/22/2024 08:00PM ET   
340  32707721       SG/G/UTIL    3000  HOU@NOP 02/22/2024 08:00PM ET   
341  32707724          C/UTIL    3000  HOU@NOP 02/22/2024 08:00PM ET   

    TeamAbbrev  AvgPointsPerGame  
0          DAL             65.18  
1          DEN             59.82  
2          PHI             63.90  
3          OKC             53.66  
4          PHX             49.34  
..         ...               ...  
337        HOU              7.11  
338        HOU              6.39  
339        HOU              5.58  
340        HOU              5.88  
341        HOU              0.00  

[342 rows x 9 columns]

In [ ]:
player_stats

Rk               PlayerName Pos  Age   Tm   G  GS    MP   FG   FGA  ...  \
0      1  Shai Gilgeous-Alexander  PG   25  OKC  53  53  1831  580  1063  ...   
1      2              Luka Dončić  PG   24  DAL  47  47  1760  542  1102  ...   
2      3            Jalen Brunson  PG   27  NYK  52  52  1865  511  1059  ...   
3      4             Jayson Tatum  PF   25  BOS  52  52  1860  481  1012  ...   
4      5             Nikola Jokić   C   28  DEN  53  53  1793  541   937  ...   
..   ...                      ...  ..  ...  ...  ..  ..   ...  ...   ...  ...   
319  320            Onuralp Bitim  SG   24  CHI   1   0     3    0     0  ...   
320  321          Malcolm Cazalon  SG   22  DET   1   0     3    0     0  ...   
321  322            Johnny Juzang  SG   22  UTA   2   0     6    0     3  ...   
322  323             Micah Potter  PF   25  UTA   2   0     7    0     1  ...   
323  324            Jason Preston  PG   24  UTA   1   0     1    0     0  ...   

     AST  STL  BLK  TOV   PF   PTS  Player-additional  SalaryPS      RosterPS  \
0    346  115   47  114  134  1649          gilgesh01     10500     PG/G/UTIL   
1    448   68   29  186   93  1606          doncilu01     12100     PG/G/UTIL   
2    336   48   10  123  105  1434          brunsja01      9300     PG/G/UTIL   
3    249   52   31  130  104  1411          tatumja01      9900  SF/PF/F/UTIL   
4    474   61   49  156  134  1382          jokicni01     11600        C/UTIL   
..   ...  ...  ...  ...  ...   ...                ...       ...           ...   
319    0    0    0    0    0     0          bitimon01      3000     SG/G/UTIL   
320    0    0    0    0    0     0          cazalma01      3000     SF/F/UTIL   
321    1    0    0    0    0     0          juzanjo01      3000     SG/G/UTIL   
322    0    0    0    0    0     0          pottemi01      3000        C/UTIL   
323    0    0    0    1    0     0          prestja01      3000     PG/G/UTIL   

                        GameInfoPS  
0    LAC@OKC 02/22/2024 08:00PM ET  
1    PHX@DAL 02/22/2024 07:30PM ET  
2    NYK@PHI 02/22/2024 07:00PM ET  
3    BOS@CHI 02/22/2024 08:00PM ET  
4    WAS@DEN 02/22/2024 09:00PM ET  
..                             ...  
319  BOS@CHI 02/22/2024 08:00PM ET  
320  DET@IND 02/22/2024 07:00PM ET  
321  CHA@UTA 02/22/2024 09:00PM ET  
322  CHA@UTA 02/22/2024 09:00PM ET  
323  CHA@UTA 02/22/2024 09:00PM ET  

[324 rows x 34 columns]

In [ ]:
feb22_stats

Rk                   Player       Date     Age Team Unnamed: 5  Opp  \
0      1              Luka Dončić 2024-02-22  24-359  DAL        NaN  PHO   
1      2             Devin Booker 2024-02-22  27-115  PHO          @  DAL   
2      3             Tyrese Maxey 2024-02-22  23-110  PHI        NaN  NYK   
3      4  Shai Gilgeous-Alexander 2024-02-22  25-225  OKC        NaN  LAC   
4      5               Kyle Kuzma 2024-02-22  28-213  WAS          @  DEN   
..   ...                      ...        ...     ...  ...        ...  ...   
157  158           Daniel Gafford 2024-02-22  25-144  DAL        NaN  PHO   
158  159        Derrick Jones Jr. 2024-02-22  27-007  DAL        NaN  PHO   
159  160            Miles McBride 2024-02-22  23-167  NYK          @  PHI   
160  161              Sam Merrill 2024-02-22  27-283  CLE        NaN  ORL   
161  162           Vasilije Micić 2024-02-22  30-040  CHO          @  UTA   

        Result Starter  MP  ...  BLK  TOV  PF  PTS  Pos.  SalaryFeb  \
0    W 123-113       *  38  ...    0    3   4   41   G-F      12100   
1    L 113-123       *  39  ...    0    4   2   35     G       9800   
2     L 96-110       *  40  ...    0    4   1   35     G       9100   
3    W 129-107       *  35  ...    2    2   1   31     G      10500   
4    L 110-130       *  31  ...    0    3   4   31     F       7400   
..         ...     ...  ..  ...  ...  ...  ..  ...   ...        ...   
157  W 123-113     NaN  17  ...    1    1   4    4   C-F       6000   
158  W 123-113     NaN  18  ...    1    0   0    4     F       3600   
159   W 110-96     NaN  17  ...    0    0   0    4     G       4200   
160  L 109-116     NaN  12  ...    0    0   0    4     G       3000   
161  W 115-107     NaN  20  ...    0    2   3    4     G       3900   

        RosterPos  Double-Double  Triple-Double                       GameInfo  
0       PG/G/UTIL            Yes             No  PHX@DAL 02/22/2024 07:30PM ET  
1    PG/SG/G/UTIL             No             No  PHX@DAL 02/22/2024 07:30PM ET  
2       PG/G/UTIL             No             No  NYK@PHI 02/22/2024 07:00PM ET  
3       PG/G/UTIL             No             No  LAC@OKC 02/22/2024 08:00PM ET  
4     PF/C/F/UTIL            Yes             No  WAS@DEN 02/22/2024 09:00PM ET  
..            ...            ...            ...                            ...  
157        C/UTIL             No             No  PHX@DAL 02/22/2024 07:30PM ET  
158  SF/PF/F/UTIL             No             No  PHX@DAL 02/22/2024 07:30PM ET  
159  PG/SG/G/UTIL             No             No  NYK@PHI 02/22/2024 07:00PM ET  
160     SG/G/UTIL             No             No  ORL@CLE 02/22/2024 07:00PM ET  
161  PG/SG/G/UTIL             No             No  CHA@UTA 02/22/2024 09:00PM ET  

[162 rows x 37 columns]

**Preliminary Model Setup (Only 8 Player Constraint)**

To initiate the process, I've created an empty model to grasp the mechanics of the problem, focusing solely on the 8-player constraint. Subsequently, I will integrate additional constraints and fully address the problem.

To ensure accessibility, I've stored the Draft Kings (DK) scoring rules from figure 1 in a dataframe since the excel file lacked this information. Following that, I established a dictionary for the scoring rules to facilitate easier reference during the analysis.






In [ ]:
# Scoring rules DataFrame
scoring_rules = pd.DataFrame({
    'Category': ['PTS', '3P', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'Double-Double', 'Triple-Double'],
    'Points': [1, 0.5, 1.25, 1.5, 2, 2, -0.5, 1.5, 3]
})
scoring_rules

Category  Points
0            PTS    1.00
1             3P    0.50
2            TRB    1.25
3            AST    1.50
4            STL    2.00
5            BLK    2.00
6            TOV   -0.50
7  Double-Double    1.50
8  Triple-Double    3.00

In [ ]:
# Create a dictionary for easier lookup
scoring_dict = dict(zip(scoring_rules['Category'], scoring_rules['Points']))

**Model Setup and Decison Variables**

With all the requisite data in hand for solving problem 1, we can proceed to establish the model and define the decision variables. These variables, denoting the players selected, will be binary, with a value of 1 indicating selection and 0 indicating non-selection.






In [ ]:
# Define the Pyomo Concrete Model
model = pe.ConcreteModel()

In [ ]:
# Define decision variables
model.players = pe.Set(initialize=feb22_stats.index)
model.x = pe.Var(model.players, domain=pe.Binary)

**Define the Objective Function**

Our objective function aims to maximize the accumulation of DK points as outlined in the scoring rules. We aim to select players who can maximize these points based on their performance in the games they participated in on Feb 22.

In [ ]:
# Define the objective function using the scoring dictionary
model.obj = pe.Objective(
    expr=sum(
        scoring_dict[category] * feb22_stats[category][i] * model.x[i] if category not in ['Double-Double', 'Triple-Double']
        else scoring_dict[category] * (feb22_stats[category][i] == 'Yes') * model.x[i]
        for category in scoring_dict.keys()
        for i in model.players
    ),
    sense=pe.maximize
)
model.obj.pprint()

obj : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : maximize : 41.0*x[0] + 35.0*x[1] + 35.0*x[2] + 31.0*x[3] + 31.0*x[4] + 30.0*x[5] + 29.0*x[6] + 28.0*x[7] + 28.0*x[8] + 27.0*x[9] + 26.0*x[10] + 25.0*x[11] + 25.0*x[12] + 25.0*x[13] + 24.0*x[14] + 24.0*x[15] + 23.0*x[16] + 22.0*x[17] + 22.0*x[18] + 22.0*x[19] + 22.0*x[20] + 22.0*x[21] + 21.0*x[22] + 21.0*x[23] + 21.0*x[24] + 21.0*x[25] + 21.0*x[26] + 20.0*x[27] + 20.0*x[28] + 20.0*x[29] + 20.0*x[30] + 20.0*x[31] + 19.0*x[32] + 19.0*x[33] + 19.0*x[34] + 19.0*x[35] + 18.0*x[36] + 18.0*x[37] + 18.0*x[38] + 18.0*x[39] + 18.0*x[40] + 18.0*x[41] + 18.0*x[42] + 18.0*x[43] + 18.0*x[44] + 17.0*x[45] + 17.0*x[46] + 17.0*x[47] + 17.0*x[48] + 17.0*x[49] + 17.0*x[50] + 16.0*x[51] + 16.0*x[52] + 16.0*x[53] + 16.0*x[54] + 16.0*x[55] + 15.0*x[56] + 15.0*x[57] + 15.0*x[58] + 14.0*x[59] + 14.0*x[60] + 14.0*x[61] + 14.0*x[62] + 14.0*x[63] + 14.0*x[64] + 14.0*x[65] + 14.0*x[66] + 14.0*x[67] + 13.0*x[68] + 13

**Define Constraints**

At this stage, we'll begin by establishing the initial constraint of having precisely 8 players, as we're solely focused on setting up the model to grasp the mechanics of the problem.






In [ ]:
# Define the one constraint
model.total_players = pe.Constraint(expr=sum(model.x[i] for i in model.players) == 8)

**Solve the Model**

In [ ]:
# Solve the simple optimization model
opt = pe.SolverFactory('glpk')
result = opt.solve(model)
print(result.solver.status, result.solver.termination_condition)

ok optimal


**See Output**

Let's look at the output to see what the model looks like without any constraints except for picking the 8 players.

In [ ]:
# Display the selected lineup and total points
selected_lineup = [i for i in model.players if pe.value(model.x[i]) == 1]
total_points = sum(
    scoring_dict[category] * feb22_stats[category][i] if category not in ['Double-Double', 'Triple-Double']
    else scoring_dict[category] * (feb22_stats[category][i] == 'Yes')
    for category in scoring_dict.keys()
    for i in selected_lineup
)

print("Total DK Points: {:.2f}".format(total_points))

Total DK Points: 473.50


In [ ]:
# Extract selected lineup information
selected_lineup_data = feb22_stats.loc[selected_lineup]

# Display the selected lineup in a table
output_table = selected_lineup_data[['Player', 'PTS', '3P', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'Double-Double', 'Triple-Double', 'SalaryFeb']].copy()
output_table['DK_Points'] = sum(
    scoring_dict[category] * selected_lineup_data[category] if category not in ['Double-Double', 'Triple-Double']
    else scoring_dict[category] * (selected_lineup_data[category] == 'Yes')
    for category in scoring_dict.keys()
)

# Display the table
output_table


Player  PTS  3P  TRB  AST  STL  BLK  TOV Double-Double  \
0               Luka Dončić   41   6    9   11    3    0    3           Yes   
1              Devin Booker   35   2    5    8    1    0    4            No   
3   Shai Gilgeous-Alexander   31   3    4    8    1    2    2            No   
4                Kyle Kuzma   31   6   12    3    3    0    3           Yes   
5           Cade Cunningham   30   4    5    8    1    0    3            No   
10            Miles Bridges   26   4   14    4    0    0    1           Yes   
11        Tyrese Haliburton   25   4    2   13    2    0    5           Yes   
25             Nikola Jokić   21   0   19   15    1    0    3           Yes   

   Triple-Double  SalaryFeb  DK_Points  
0             No      12100      77.75  
1             No       9800      54.25  
3             No      10500      54.50  
4             No       7400      59.50  
5             No       7600      50.75  
10            No       8900      52.50  
11            No       9400      52.00  
25           Yes      11600      72.25

Although our lineup achieves a total DK points of 473.50 from the players selected for our roster, it does so without adhering to any constraints except for the limit of 8 players. Despite maximizing DK points from the Feb 22 games, this lineup blatantly disregards all other constraints.






## Problem 2

So now that we understand how the problem works and what we're solving for, let's fully solve problem 2.

**Model Setup and Decison Variables**

In [ ]:
# Define the Pyomo Concrete Model
model = pe.ConcreteModel()

In [ ]:
# Define decision variables
model.players = pe.Set(initialize=feb22_stats.index)
model.x = pe.Var(model.players, domain=pe.Binary)
for DV in model.component_objects(pe.Var):
    DV.pprint()

x : Size=162, Index=players
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :     1 : False :  True : Binary
      1 :     0 :  None :     1 : False :  True : Binary
      2 :     0 :  None :     1 : False :  True : Binary
      3 :     0 :  None :     1 : False :  True : Binary
      4 :     0 :  None :     1 : False :  True : Binary
      5 :     0 :  None :     1 : False :  True : Binary
      6 :     0 :  None :     1 : False :  True : Binary
      7 :     0 :  None :     1 : False :  True : Binary
      8 :     0 :  None :     1 : False :  True : Binary
      9 :     0 :  None :     1 : False :  True : Binary
     10 :     0 :  None :     1 : False :  True : Binary
     11 :     0 :  None :     1 : False :  True : Binary
     12 :     0 :  None :     1 : False :  True : Binary
     13 :     0 :  None :     1 : False :  True : Binary
     14 :     0 :  None :     1 : False :  True : Binary
     15 :     0 :  None :     1 : False :  True : Binary
   

In [ ]:
# Define the objective function using the scoring dictionary
model.obj = pe.Objective(
    expr=sum(
        scoring_dict[category] * feb22_stats[category][i] * model.x[i] if category not in ['Double-Double', 'Triple-Double']
        else scoring_dict[category] * (feb22_stats[category][i] == 'Yes') * model.x[i]
        for category in scoring_dict.keys()
        for i in model.players
    ),
    sense=pe.maximize
)
model.obj.pprint()

obj : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : maximize : 41.0*x[0] + 35.0*x[1] + 35.0*x[2] + 31.0*x[3] + 31.0*x[4] + 30.0*x[5] + 29.0*x[6] + 28.0*x[7] + 28.0*x[8] + 27.0*x[9] + 26.0*x[10] + 25.0*x[11] + 25.0*x[12] + 25.0*x[13] + 24.0*x[14] + 24.0*x[15] + 23.0*x[16] + 22.0*x[17] + 22.0*x[18] + 22.0*x[19] + 22.0*x[20] + 22.0*x[21] + 21.0*x[22] + 21.0*x[23] + 21.0*x[24] + 21.0*x[25] + 21.0*x[26] + 20.0*x[27] + 20.0*x[28] + 20.0*x[29] + 20.0*x[30] + 20.0*x[31] + 19.0*x[32] + 19.0*x[33] + 19.0*x[34] + 19.0*x[35] + 18.0*x[36] + 18.0*x[37] + 18.0*x[38] + 18.0*x[39] + 18.0*x[40] + 18.0*x[41] + 18.0*x[42] + 18.0*x[43] + 18.0*x[44] + 17.0*x[45] + 17.0*x[46] + 17.0*x[47] + 17.0*x[48] + 17.0*x[49] + 17.0*x[50] + 16.0*x[51] + 16.0*x[52] + 16.0*x[53] + 16.0*x[54] + 16.0*x[55] + 15.0*x[56] + 15.0*x[57] + 15.0*x[58] + 14.0*x[59] + 14.0*x[60] + 14.0*x[61] + 14.0*x[62] + 14.0*x[63] + 14.0*x[64] + 14.0*x[65] + 14.0*x[66] + 14.0*x[67] + 13.0*x[68] + 13

**Define Constraints**


In [ ]:
# Define constraints
model.total_players = pe.Constraint(expr=sum(model.x[i] for i in model.players) == 8)
model.total_salary = pe.Constraint(expr=sum(feb22_stats['SalaryFeb'][i] * model.x[i] for i in model.players) <= 50000)
model.games_constraint = pe.Constraint(expr=sum(model.x[i] for i in model.players if feb22_stats['GameInfo'][i] != feb22_stats['GameInfo'][selected_lineup[0]]) >= 1)
# Constraint: At least one Point Guard (PG)
model.PG = pe.Constraint(expr=sum(model.x[i] for i in model.players if 'PG' in feb22_stats.loc[i, 'RosterPos'].split('/')) >= 2)
# Constraint: At least one Shooting Guard (SG)
model.SG = pe.Constraint(expr=sum(model.x[i] for i in model.players if 'SG' in feb22_stats.loc[i, 'RosterPos'].split('/')) >= 2)
# Constraint: At least one Small Forward (SF)
model.SF = pe.Constraint(expr=sum(model.x[i] for i in model.players if 'SF' in feb22_stats.loc[i, 'RosterPos'].split('/')) >= 2)
# Constraint: At least one Power Forward (PF)
model.PF = pe.Constraint(expr=sum(model.x[i] for i in model.players if 'PF' in feb22_stats.loc[i, 'RosterPos'].split('/')) >= 2)
# Constraint: At least one Center (C)
model.C = pe.Constraint(expr=sum(model.x[i] for i in model.players if 'C' in feb22_stats.loc[i, 'RosterPos'].split('/')) >= 1)
#Utility Constraint - No specific constraint needed since the utility position allows for any of the 5 positions, and it is implicitly fulfilled by the other constraints.


**Solve the Model to see if it's Optimal**

In [ ]:
opt = pe.SolverFactory('glpk')
result = opt.solve(model)
print(result.solver.status, result.solver.termination_condition)

ok optimal


What's our new total DK points given our all our constraints for problem 2?

In [ ]:
# Display the selected lineup and total points
selected_lineup = [i for i in model.players if pe.value(model.x[i]) == 1]
total_points = sum(
    scoring_dict[category] * feb22_stats[category][i] if category not in ['Double-Double', 'Triple-Double']
    else scoring_dict[category] * (feb22_stats[category][i] == 'Yes')
    for category in scoring_dict.keys()
    for i in selected_lineup
)

print("Total DK Points: {:.2f}".format(total_points))

Total DK Points: 386.75


**Problem 2 Lineup**


Now that we have derived the optimal solution, let's analyze which players stand out as the best performers considering all imposed constraints.






In [ ]:
# Extract selected lineup information
selected_lineup_data = feb22_stats.loc[selected_lineup]

# Display the selected lineup in a table
output_table = selected_lineup_data[['Player', 'RosterPos', 'GameInfo', 'PTS', '3P', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'Double-Double', 'Triple-Double', 'SalaryFeb']].copy()
output_table['DK_Points'] = sum(
    scoring_dict[category] * selected_lineup_data[category] if category not in ['Double-Double', 'Triple-Double']
    else scoring_dict[category] * (selected_lineup_data[category] == 'Yes')
    for category in scoring_dict.keys()
)

# Display the table
output_table

Player     RosterPos                       GameInfo  PTS  3P  \
0         Luka Dončić     PG/G/UTIL  PHX@DAL 02/22/2024 07:30PM ET   41   6   
4          Kyle Kuzma   PF/C/F/UTIL  WAS@DEN 02/22/2024 09:00PM ET   31   6   
8       Derrick White  PG/SG/G/UTIL  BOS@CHI 02/22/2024 08:00PM ET   28   5   
13     Gary Trent Jr.  PG/SG/G/UTIL  BKN@TOR 02/22/2024 07:00PM ET   25   5   
30  Jonas Valančiūnas        C/UTIL  HOU@NOP 02/22/2024 08:00PM ET   20   2   
48      Herbert Jones  SF/PF/F/UTIL  HOU@NOP 02/22/2024 08:00PM ET   17   3   
54     T.J. McConnell     PG/G/UTIL  DET@IND 02/22/2024 07:00PM ET   16   2   
92     Ausar Thompson  SF/PF/F/UTIL  DET@IND 02/22/2024 07:00PM ET   11   0   

    TRB  AST  STL  BLK  TOV Double-Double Triple-Double  SalaryFeb  DK_Points  
0     9   11    3    0    3           Yes            No      12100      77.75  
4    12    3    3    0    3           Yes            No       7400      59.50  
8     3    5    2    3    2            No            No       6400      50.75  
13    1    2    1    0    0            No            No       4000      33.75  
30   14    3    1    0    1           Yes            No       6400      46.00  
48    5    4    7    0    1            No            No       4600      44.25  
54    6    5    2    1    1            No            No       4100      37.50  
92    7    5    2    3    0            No            No       5000      37.25

What did the players' total salary come out to?

In [ ]:
# Calculate total salary for all players in the lineup
total_salary = sum(output_table['SalaryFeb'])

# Display the total salary
print("Total Salary for Selected Lineup: {:.2f}".format(total_salary))

Total Salary for Selected Lineup: 50000.00


What is the players total average points


In [ ]:
total_average_points=output_table['PTS'].mean()

average_points_per_nba= feb22_stats['PTS'].mean()

print("Average Points for Every Game on Feb 22: {:.2f}".format(average_points_per_nba))

print("Average Points for Lineup chose on Feb 22: {:.2f}".format(total_average_points))


Average Points for Every Game on Feb 22: 13.62
Average Points for Lineup chose on Feb 22: 23.62


**Problem 2 Summary**

After considering all constraints, the optimal DK points achieved from the Feb 22 games amounted to 386.75. This value notably decreased compared to the initial unrestricted scenario where only the selection of 8 players was mandated. The chosen lineup effectively maximizes the total DK points from the games while adhering to the $50,000 combined salary limit and fulfilling all positional constraints. Additionally, the lineup ensures players participated in at least two different games.

Among the selected players, Luka Dončić, Kyle Kuzma, and Derrick White emerged as the top performers in terms of DK points. For detailed information regarding each player's DK points in their respective games, refer to the far right column labeled DK_Points.






##Problem 1

To initiate the solution process for problem 1, I began by compiling the Draft Kings (DK) scoring rules outlined in Figure 1 into a dataframe. It's worth noting that the Double-Double and Triple-Double scoring rules were excluded in this dataframe, as we are assuming a lack of knowledge regarding player performance in the Feb 22 games. Further details on the approach taken to tackle this problem will be discussed in the objective function setup section.






In [ ]:
# Scoring rules DataFrame
scoring_rules_1 = pd.DataFrame({
    'Category': ['PTS', '3P', 'TRB', 'AST', 'STL', 'BLK', 'TOV'],
    'Points': [1, 0.5, 1.25, 1.5, 2, 2, -0.5]
})
scoring_rules_1

Category  Points
0      PTS    1.00
1       3P    0.50
2      TRB    1.25
3      AST    1.50
4      STL    2.00
5      BLK    2.00
6      TOV   -0.50

In [ ]:
# Create a dictionary for easier lookup
scoring_dict_1 = dict(zip(scoring_rules_1['Category'], scoring_rules_1['Points']))

**Model Setup and Decison Variables**


In [ ]:
# Define the Pyomo Concrete Model for Problem 1
model_1 = pe.ConcreteModel()

In [ ]:
# Define decision variables for Problem 1
model_1.players = pe.Set(initialize=player_stats.index)
model_1.x = pe.Var(model_1.players, domain=pe.Binary)
for DV in model_1.component_objects(pe.Var):
    DV.pprint()

x : Size=324, Index=players
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :     1 : False :  True : Binary
      1 :     0 :  None :     1 : False :  True : Binary
      2 :     0 :  None :     1 : False :  True : Binary
      3 :     0 :  None :     1 : False :  True : Binary
      4 :     0 :  None :     1 : False :  True : Binary
      5 :     0 :  None :     1 : False :  True : Binary
      6 :     0 :  None :     1 : False :  True : Binary
      7 :     0 :  None :     1 : False :  True : Binary
      8 :     0 :  None :     1 : False :  True : Binary
      9 :     0 :  None :     1 : False :  True : Binary
     10 :     0 :  None :     1 : False :  True : Binary
     11 :     0 :  None :     1 : False :  True : Binary
     12 :     0 :  None :     1 : False :  True : Binary
     13 :     0 :  None :     1 : False :  True : Binary
     14 :     0 :  None :     1 : False :  True : Binary
     15 :     0 :  None :     1 : False :  True : Binary
   

In addressing problem 1, the objective is to construct the optimal starting lineup by evaluating player performance across all DraftKings (DK) categories up to the specified date (Feb 22). This entails considering every DK category except for double-doubles and triple-doubles, as these are bonuses applied per game, while the available PlayerStats data covers all NBA games up to Feb 22. The selection process is guided by identifying players who have excelled within the given constraints and additional criteria:

Exclusion of players from the two lowest-performing NBA teams at present (Detroit Pistons - DET or Washington Wizards - WAS).
Enforcing a minimum Field Goal Percentage (FG%) threshold of above .45 (45%), a benchmark that selects players with an FG%  that meets the league average,
Imposing a minimum requirement of 1000 minutes played throughout the season. Given the average minutes per player per game in the NBA (approximately 20 minutes) multiplied by the number of games played up to Feb 22 (56), this criterion ensures the selection of players who have received substantial playing time, enhancing their potential to accumulate points in games.





In [ ]:
# Define the objective function using the scoring dictionary for Problem 1
model_1.obj = pe.Objective(
    expr=sum(
        scoring_dict_1[category] * player_stats[category][i] * model_1.x[i]
        for category in scoring_dict_1.keys()
        for i in model_1.players
    ),
    sense=pe.maximize
)
model_1.obj.pprint()

obj : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : maximize : 1649.0*x[0] + 1606.0*x[1] + 1434.0*x[2] + 1411.0*x[3] + 1382.0*x[4] + 1353.0*x[5] + 1265.0*x[6] + 1259.0*x[7] + 1249.0*x[8] + 1237.0*x[9] + 1202.0*x[10] + 1201.0*x[11] + 1186.0*x[12] + 1170.0*x[13] + 1156.0*x[14] + 1136.0*x[15] + 1123.0*x[16] + 1122.0*x[17] + 1103.0*x[18] + 1102.0*x[19] + 1099.0*x[20] + 1093.0*x[21] + 1077.0*x[22] + 1055.0*x[23] + 991.0*x[24] + 986.0*x[25] + 972.0*x[26] + 966.0*x[27] + 965.0*x[28] + 952.0*x[29] + 939.0*x[30] + 926.0*x[31] + 921.0*x[32] + 903.0*x[33] + 899.0*x[34] + 896.0*x[35] + 894.0*x[36] + 873.0*x[37] + 864.0*x[38] + 861.0*x[39] + 851.0*x[40] + 840.0*x[41] + 838.0*x[42] + 820.0*x[43] + 813.0*x[44] + 809.0*x[45] + 798.0*x[46] + 794.0*x[47] + 781.0*x[48] + 780.0*x[49] + 775.0*x[50] + 774.0*x[51] + 763.0*x[52] + 755.0*x[53] + 747.0*x[54] + 743.0*x[55] + 737.0*x[56] + 732.0*x[57] + 728.0*x[58] + 712.0*x[59] + 710.0*x[60] + 681.0*x[61] + 677.0*x

**Define Constraints**

Let's outline the constraints provided in the problem statement as well as the additional constraints introduced in our discussion:

In [ ]:
# Define constraints for Problem 1

# We need exactly 8 players
model_1.total_players = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players) == 8)
# The total salary for all the payers has to be less than or equal to 50,000
model_1.total_salary = pe.Constraint(expr=sum(player_stats['SalaryPS'][i] * model_1.x[i] for i in model_1.players) <= 50000)
# We need players from at least 2 games
model_1.games_constraint = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats['GameInfoPS'][i] != player_stats['GameInfoPS'][selected_lineup[0]]) >= 1)
# We need players in these positions
model_1.PG = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if 'PG' in player_stats.loc[i, 'RosterPS'].split('/')) >= 2)
model_1.SG = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if 'SG' in player_stats.loc[i, 'RosterPS'].split('/')) >= 2)
model_1.SF = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if 'SF' in player_stats.loc[i, 'RosterPS'].split('/')) >= 2)
model_1.PF = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if 'PF' in player_stats.loc[i, 'RosterPS'].split('/')) >= 2)
model_1.C = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if 'C' in player_stats.loc[i, 'RosterPS'].split('/')) >= 1)

# New Additional Constraints for Problem 1
# FG% more than .45
model_1.FG_percentage = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats.loc[i, 'FG%'] >= 0.45) >= 8)
# No players from DET or WAS
model_1.bad_teams_constraint = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats['Tm'][i] in ['DET', 'WAS']) == 0)
# Players must have played for more than 1,000 minutes so far this season
model_1.minutes_constraint = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats.loc[i, 'MP'] >= 1000) >= 8)



This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().
This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


**Solve the Model to see if it's Optimal**

In [ ]:
opt = pe.SolverFactory('glpk')
result = opt.solve(model_1)
print(result.solver.status, result.solver.termination_condition)

ok optimal


What is the cumulative DK Points tally for the selected players throughout the season, considering the newly imposed constraints?

In [ ]:
# Display the selected lineup and total points for Problem 1
selected_lineup_1 = [i for i in model_1.players if pe.value(model_1.x[i]) == 1]
total_points_1 = sum(
    scoring_dict_1[category] * player_stats[category][i] if category not in ['Double-Double', 'Triple-Double']
    else scoring_dict_1[category] * (player_stats[category][i] == 'Yes')
    for category in scoring_dict_1.keys()
    for i in selected_lineup_1
)

print("Total DK Points for the season before Feb 22 games: {:.2f}".format(total_points_1))


Total DK Points for the season before Feb 22 games: 13596.25


**Problem 1 Lineup**

In [ ]:
# Extract selected lineup information for Problem 1
selected_lineup_data_1 = player_stats.loc[selected_lineup_1]

# Display the selected lineup in a table for Problem 1
output_table_1 = selected_lineup_data_1[['PlayerName', 'Tm', 'MP', 'RosterPS', 'SalaryPS', 'GameInfoPS', 'FG%', 'PTS', '3P', 'TRB', 'AST', 'STL', 'BLK', 'TOV']].copy()
output_table_1['DK_Points'] = sum(
    scoring_dict_1[category] * selected_lineup_data_1[category]
    for category in scoring_dict_1.keys()
)

# Display the table for Problem 1
output_table_1



PlayerName   Tm    MP        RosterPS  SalaryPS  \
6        Paolo Banchero  ORL  1923       PF/F/UTIL      8400   
10        DeMar DeRozan  CHI  1996    SF/PF/F/UTIL      7600   
18       Scottie Barnes  TOR  1935  PG/SF/F/G/UTIL      8800   
33        Chet Holmgren  OKC  1624          C/UTIL      7300   
53    Jonas Valančiūnas  NOP  1403          C/UTIL      6400   
57   Bennedict Mathurin  IND  1325  SG/SF/F/G/UTIL      4500   
78           Obi Toppin  IND  1259       PF/F/UTIL      3700   
138       Cason Wallace  OKC  1120    PG/SG/G/UTIL      3200   

                        GameInfoPS    FG%   PTS  3P  TRB  AST  STL  BLK  TOV  \
6    ORL@CLE 02/22/2024 07:00PM ET  0.462  1265  85  377  289   58   29  165   
10   BOS@CHI 02/22/2024 08:00PM ET  0.472  1202  49  216  278   61   33   80   
18   BKN@TOR 02/22/2024 07:00PM ET  0.473  1103  98  448  328   67   80  149   
33   LAC@OKC 02/22/2024 08:00PM ET  0.535   903  90  409  144   38  143   88   
53   HOU@NOP 02/22/2024 08:00PM ET  0.573   755  30  517  129   24   57   76   
57   DET@IND 02/22/2024 07:00PM ET  0.450   732  64  199   98   28   11   86   
78   DET@IND 02/22/2024 07:00PM ET  0.568   597  71  218   87   29   28   48   
138  LAC@OKC 02/22/2024 08:00PM ET  0.504   367  63  117   72   37   23   25   

     DK_Points  
6      2303.75  
10     2061.50  
18     2423.50  
33     1993.25  
53     1733.75  
57     1194.75  
78     1125.50  
138     760.25

What did the players' total salary come out to?

In [ ]:
# Calculate total salary for the selected lineup
total_salary = output_table_1['SalaryPS'].sum()

# Display the total salary
print("Total Salary for Selected Lineup: {:.2f}".format(total_salary))

Total Salary for Selected Lineup: 49900.00


What's the players total avg per game

In [ ]:
total_average_points_chose=(player_stats.loc[selected_lineup_1,"PTS"]/player_stats.loc[selected_lineup_1,"G"]).mean()

average_points_chose=(player_stats["PTS"]/player_stats["G"]).mean()


print("Average Points for Every Player in NBA {:.2f}".format(average_points_chose))

print("Average Points for my Optimal Lineup {:.2f}".format(total_average_points_chose))


Average Points for Every Player in NBA 8.97
Average Points for my Optimal Lineup 15.96


**Problem 1 Summary**


Reviewing the results, it's evident that Paolo Banchero, Demar DeRozan, and Scottie Barnes emerge as the top three performers this season, considering our specified constraints. The introduction of new criteria, including a minimum FG% of .45 and a minimum game time of 1000 minutes, alongside the exclusion of players from the Washington Wizards or Detroit Pistons, significantly influenced player selection.

In light of these constraints, these players represent the optimal choices within the given parameters. Notably, the total salary for this lineup amounted to $49,900,

effectively utilizing the allocated $50,000 limit. For a comprehensive view of player performance within our constraints, refer to the complete player list in the far right column of the table above.










##**Final Summary**

As evident from the analysis, accurately predicting player performance without comprehensive data, such as the Feb 22 game results, poses a challenge. Only one player, Jonas Valančiūnas, was retained from the previous lineup, indicating limitations in the predictive capability of the second model compared to the first.

While the second model may not match the accuracy of the first due to the absence of crucial data, it's encouraging that at least one player overlapped between the lineups. Reflecting on this process, future iterations could explore additional factors such as opponents' strength and home versus away game dynamics to enhance predictive power.

However, it's important to acknowledge that introducing more constraints may reduce the optimal value. Despite this potential trade-off, it would be intriguing to investigate whether a refined lineup can be achieved by incorporating such nuanced considerations, even at the expense of a slightly lower optimal value.







##Appendix 1 - Ask your favorite GAI what it would do

I asked  CHAT GPT to Solve this problem I gave it all the details such as the Draft Kings Scoring Table for Fantansy Basketball, Data Dictionary, constraints,and the other neccesary requirements to solve this problem.

Here is the code that it produced I have it commented it out because it did not run. This problem is not the type of problem to just prompt Chat GPT to solve the whole thing however, it can help while going with the step by step process while solving this problem

In [ ]:

# # Load data from Excel workbook
# excel_file = pd.ExcelFile('NBA_Project.xlsx')
# dk_salaries = excel_file.parse('DKSalaries')
# game_data = excel_file.parse('ActualGameData')  # Replace XXX with the actual game data sheet name
# player_stats = excel_file.parse('PlayerStats')

# # Ensure the dataframes have the same index
# dk_salaries = dk_salaries.reset_index(drop=True)
# game_data = game_data.reset_index(drop=True)
# player_stats = player_stats.reset_index(drop=True)

# # Define the Pyomo model
# model = ConcreteModel()

# # Define decision variables for each player (binary variables)
# model.players = Var(dk_salaries.index, within=Binary)

# # Objective function: maximize total average points per game from dk_salaries
# model.objective = Objective(expr=sum(
#     dk_salaries['AvgPointsPerGame'].iloc[i] * model.players[i]
#     for i in dk_salaries.index),
#     sense=maximize)

# # Constraints
# model.total_players = Constraint(expr=sum(model.players[i] for i in dk_salaries.index) == 8)
# # Constraint: Total salary should not exceed $50,000
# model.total_salary = Constraint(expr=sum(row['Salary'] * model.players[i] for i, row in dk_salaries.iterrows()) <= 50000)

# unique_games = set(dk_salaries['Game Info'].str.split(',').explode().str.strip())

# # Create a binary variable for each game
# model.games_binary = Var(unique_games, within=Binary)

# # Constraint: Players must be from at least two different games
# model.min_games = Constraint(expr=sum(model.games_binary[game] for game in unique_games) >= 2)
# model.link_players_games = ConstraintList()

# for i in dk_salaries.index:
#     game_info = dk_salaries.at[i, 'Game Info'].split(',')
#     for game in game_info:
#         game = game.strip()
#         model.link_players_games.add(model.players[i] <= model.games_binary[game])



# model.positions = [Constraint(expr=sum(model.players[i] for i in dk_salaries[dk_salaries['Roster Position'] == pos].index) == 1)
#                    for pos in ['PG', 'SG', 'SF', 'PF', 'C', 'G', 'F', 'Util']]
# # ... (Add any other constraints based on your specific preferences or strategies)

# # Solve the optimization problem
# solver = SolverFactory('glpk')
# solver.solve(model)

# # Display the selected lineup
# selected_lineup = dk_salaries.iloc[[i for i in dk_salaries.index if model.players[i].value == 1]]
# print(selected_lineup[['Name', 'Roster Position', 'Salary']])


##Appendix 2 - Any documentation of the use of GAI

I used AI in help


*   Generating and adjusting my additional constraints for problem 1
*   Helping display my results
*  Creating a dictionaries  for easier lookup.

Throughout this project, AI played a pivotal role by aiding me in generating code solutions that would have otherwise remained elusive. Its assistance proved invaluable, particularly during troubleshooting sessions where I encountered stubborn errors. By presenting a range of code options, AI not only broadened my understanding of different approaches but also steered me towards optimal solutions. While AI's involvement may have been modest in terms of overall project scope, its impact was profound, saving time and empowering me to unravel complex code concepts with ease.






Here is prompt to create those additonal constraints for problem 1:


Create constraints for the optimization model named model_1 with the following specifications:

Ensure that at least 8 players selected have a field goal percentage (FG%) of 45% or higher.
Restrict the selection of players from the teams 'DET' (Detroit Pistons) and 'WAS' (Washington Wizards) to zero.
Enforce that the total minutes played (MP) by selected players surpasses 1,000 minutes for the season.

In [ ]:
# # Constraint: Ensure at least 8 players with FG% >= 0.45 are selected
# model_1.FG_percentage = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats.loc[i, 'FG%'] >= 0.45) >= 8)

# # Constraint: No players from Detroit Pistons (DET) or Washington Wizards (WAS)
# model_1.bad_teams_constraint = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats['Tm'][i] in ['DET', 'WAS']) == 0)

# # Constraint: Players must have played for more than 1,000 minutes so far this season
# model_1.minutes_constraint = pe.Constraint(expr=sum(model_1.x[i] for i in model_1.players if player_stats.loc[i, 'MP'] >= 1000) >= 8)
